### PR TITLE
fix: update sort key in catalog before adding parquet file to catalog

### PR DESCRIPTION
**Pull request by alamb**
~Draft while I figure out why the other tests are diffing now~

Closes https://github.com/influxdata/influxdb_iox/issues/5090
May be related to https://github.com/influxdata/conductor/issues/1068

# Rationale
The ingester was updating the catalog with a new parquet file *before* it was updating the catalog with the partition's sort key. 

The querier, on the other hand, was assuming that the partition sort key was up to date for any parquet file that was in the catalog. 

This leads to very rare cases where the querier sees a parquet file that has no corresponding sort key in the catalog, which makes the querier cranky 👍 

```
thread 'IOx Query Executor Thread' panicked at 'partition sort key should be set when a parquet file exists', querier/src/chunk/mod.rs:408:14
```

# Changes
Update the partition sort key before the parquet file write in the ingester

# Testing
The following sophisticated 😆  test script
- [x] Failed regularly locally after < 150 iterations (faster when I was doing something like `cargo build` concurrently)
- [x] Completed successfully (1000 iterations)

```shell
# run the basic_multi_ingesters test in a loop
set -x
set -e

cd /Users/alamb/Software/influxdb_iox
export RUST_BACKTRACE=1
export CARGO_TARGET_DIR=/Users/alamb/Software/target-iox
export TEST_INTEGRATION=1
export TEST_INFLUXDB_IOX_CATALOG_DSN=postgresql://localhost:5432/alamb

for i in `seq 1 1000` ; do
    echo "****** AAL iteration $i"
    cargo test --test end_to_end  -- basic_multi_ingesters
    #/Users/alamb/Software/target-iox/debug/deps/end_to_end-af1a566230d46569 basic_multi_ingesters
done

echo "COMPLETE"
```

### CODE REVIEW

File: ingester/src/data.rs [hunk 1]
```diff
@@ -324,6 +311,23 @@ impl Persister for IngesterData {
                     })
                     .await
                     .expect("retry forever");
+
+                // Add the parquet file to the catalog until succeed
+                let parquet_file = iox_meta.to_parquet_file(partition_id, file_size, &md, |name| {
```
Comment (alamb) [inline ingester/src/data.rs:316]: I literally just moved the code down here

Comment (crepererum) [reply ingester/src/data.rs:316]: that looks correct and what I have expected the ingester to do (not meant as a blame but rather as an explanation of what the querier does).

File: ingester/src/data.rs [hunk 2]
```diff
@@ -326,6 +313,27 @@ impl Persister for IngesterData {
...
 
+            // Add the parquet file to the catalog until succeed
+            let parquet_file = iox_meta.to_parquet_file(partition_id, file_size, &md, |name| {
+                table_schema.columns.get(name).expect("Unknown column").id
+            });
+            Backoff::new(&self.backoff_config)
+                .retry_all_errors("add parquet file to catalog", || async {
+                    let mut repos = self.catalog.repositories().await;
+                    let parquet_file = repos.parquet_files().create(parquet_file.clone()).await?;
+                    debug!(
```
Comment (alamb) [inline ingester/src/data.rs:324]: I also improved this `debug!` call to be *after* the catalog update has happened and to contain the `parquet_file_id` and `table_id`

File: ingester/src/data.rs [hunk 3]
```diff
@@ -288,31 +288,18 @@ impl Persister for IngesterData {
...
+            // see a parquet file with an inconsistent sort key)
+            // https://github.com/influxdata/influxdb_iox/issues/5090
             if let Some(new_sort_key) = sort_key_update {
                 let sort_key = new_sort_key.to_columns().collect::<Vec<_>>();
                 debug!(
                     ?partition_id,
                     ?sort_key,
                     ?partition_id,
-                    "Adjusted sort key to be updated to the catalog"
+                    "Updated sort key in catalog"
```
Comment (crepererum) [inline ingester/src/data.rs:302]: Either move this debug line below the next statement or change the wording from `updated` to `updating`. I think moving the debug print would be more in line with what we do elsewhere (incl. your other change in this PR).

**Review by crepererum — APPROVED** [review]
Small nitpick regarding logging, otherwise looks good :+1:

File: ingester/src/data.rs (hunk 3 above)
Comment (alamb) [reply ingester/src/data.rs:302]: Good point -- I will move the `debug` statement down

**Pull request merged**